Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift

```diff
@@ -9,201 +9,1069 @@
+            }
+        }
+
+        /// Actions returned by `iteratorDeinitialized()`.
+        enum IteratorDeinitializedAction {
+            /// Indicates that the `Task` needs to be cancelled and
+            /// all upstream continuations need to be resumed with a `CancellationError`.
+            case cancelTaskAndUpstreamContinuations(
+                task: Task<Void, Never>,
+                upstreamContinuations: [CheckedContinuation<Void, Error>]
+            )
+            /// Indicates that nothing should be done.
+            case none
+        }
+
+        mutating func iteratorDeinitialized() -> IteratorDeinitializedAction {
+            switch state {
+            case .initial:
+                // An iterator needs to be initialized before it can be deinitialized.
+                preconditionFailure("Internal inconsistency")
+
+            case .merging(_, _, _, _, .some):
+                // An iterator was deinitialized while we have a suspended continuation.
+                preconditionFailure("Internal inconsistency")
+
+            case let .merging(task, _, upstreamContinuations, _, .none):
+                // The iterator was dropped which signals that the consumer is finished.
+                // We can transition to finished now and need to clean everything up.
+                state = .finished
+
+                return .cancelTaskAndUpstreamContinuations(
+                    task: task,
+                    upstreamContinuations: upstreamContinuations
+                )
+
+            case .upstreamThrew:
+                // The iterator was dropped which signals that the consumer is finished.
+                // We can transition to finished now. The cleanup already happened when we
+                // transitioned to `upstreamThrew`.
+                state = .finished
+
+                return .none
+
+            case .finished:
+                // We are already finished so there is nothing left to clean up.
+                // This is just the references dropping afterwards.
+                return .none
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        mutating func taskStarted(_ task: Task<Void, Never>) {
+            switch state {
+            case .initial:
+                // The user called `makeAsyncIterator` and we are starting the `Task`
+                // to consume the upstream sequences
+                state = .merging(
+                    task: task,
+                    buffer: .init(),
+                    upstreamContinuations: [], // This should reserve capacity in the variadic generics case
+                    upstreamsFinished: 0,
+                    downstreamContinuation: nil
+                )
+
+            case .merging, .upstreamThrew, .finished:
+                // We only a single iterator to be created so this must never happen.
+                preconditionFailure("Invalid state")
```

Comment:
it might be good to (if we decide these are useful to keep) list the state that is invalid and unify that with the same verbiage as the other fatal error cases.